src/main/java/codechicken/lib/render/shader/ShaderProgramBuilder.java

```diff
@@ -52,51 +62,130 @@ public ShaderProgram build() {
         return new ShaderProgram(shaders.values(), cacheCallback == null ? NULL_CALLBACK : cacheCallback);
     }
 
-    public static class ShaderObjectBuilder {
+    /**
+     * Created by KitsuneAlex on 18/11/21.
+     */
+    @SuppressWarnings("unchecked")
+    private static abstract class AbstractShaderObjectBuilder<B extends AbstractShaderObjectBuilder<B>> {
```

Comment:
Do we really need to split this into 3 classes, why cant `BinaryShaderObjectBuilder` just extend from `ShaderObjectBuilder` like we discussed?